numpy/_array_api/_types.py

```diff
@@ -0,0 +1,23 @@
+"""
+This file defines the types for type annotations.
+
+These names aren't part of the module namespace, but they are used in the
+annotations in the function signatures. The functions in the module are only
+valid for inputs that match the given type annotations.
+"""
+
+__all__ = ['Literal', 'Optional', 'Tuple', 'Union', 'array', 'device',
+           'dtype', 'SupportsDLPack', 'SupportsBufferProtocol', 'PyCapsule']
+
+from typing import Literal, Optional, Tuple, Union, TypeVar
+
+from . import (ndarray, int8, int16, int32, int64, uint8, uint16, uint32,
+               uint64, float32, float64)
+
+array = ndarray
+device = TypeVar('device')
+dtype = Literal[int8, int16, int32, int64, uint8, uint16,
+                uint32, uint64, float32, float64]
+SupportsDLPack = TypeVar('SupportsDLPack')
+SupportsBufferProtocol = TypeVar('SupportsBufferProtocol')
+PyCapsule = TypeVar('PyCapsule')
```

Comment:
```suggestion
device = Any
dtype = Type[Union[[int8, int16, int32, int64, uint8, uint16,
                   uint32, uint64, float32, float64]]
SupportsDLPack = Any
SupportsBufferProtocol = Any
PyCapsule = Any
```
A couple of things:
* Typevars should only be used in signatures whenever two or more parameters must/will be of the same type, _e.g._ `def func(a: PyCapsule) -> PyCapsule: ...`.
* `Literal` can only except `None`, boolean, integers and strings. It is not currently possible to annotate literal types.
* I assume `dtype` is ment to represent the a union of `np.generic` subclasses, and not instances?

Reply:
> I assume `dtype` is meant to represent the a union of `np.generic` subclasses, and not instances?

No, instances were meant here. Code using the dtypes in this module are meant to be portable between libraries. Meaning the _only_ valid way of spelling them is through these instances - no `'float'` or `'>f8' or `double` or anything else. And no subclass relationship between types.

There's some freedom in how to define what a single dtype like `int8` is - that's an implementation decision, I listed some options in https://numpy.org/neps/nep-0047-array-api-standard.html#the-dtype-objects.

> `Literal` can only except `None`, boolean, integers and strings. It is not currently possible to annotate literal types.

Indeed - getting this type annotation right is a question mark.

Reply:
> No, instances were meant here.

Ok, then they definitely have to be enclosed in `Type[...]`.

Reply:
A second question though: NEP 47 doesn't make any mention of `np.dtype` instances. Were those deliberately excluded or was this an oversight?

Reply:
That was deliberate, because those are numpy-specific. Just to make sure, you're asking if (for example) it should be possible to do:
```
np.array_api.ones((3, 2), dtype=np.float32)
```
instead of
```
np.array_api.ones((3, 2), dtype=np.array_api.float32)
```
? (those are not actually dtype instances, so you may mean something else)

Reply:
Not quite, what I'm asking is about the syntax below:
```python
dtype = np.dtype(np.float32)
np.array_api.ones((3, 2), dtype=dtype)
```

Reply:
I don't see a particular reason to support that. The current code in this PR may not raise, but it could.

That code is a little non-idiomatic anyway right? I don't think explicitly calling `np.dtype()` is normal. The way to get such instances though is:
```
dtype = x.dtype
np.array_api.ones((3, 2), dtype=dtype)
```
The NEP does discuss that pattern. It must work if `x` is an instance of the new array object. I don't currently see the need to make that work for `x` being an `np.ndarray` instance (but I could be wrong, that's the kind of thing you find out when writing real-world code).

Reply:
> ``` python
> dtype = x.dtype
> np.array_api.ones((3, 2), dtype=dtype)
> ```
> The NEP does discuss that pattern. It must work if x is an instance of the new array object.

That's the thing: the dtype attributes of both `ndarray` and `_array_api.Array` currently return `np.dtype` instances, 
so neither of them can currently be used as `x` in this example.

Reply:
> That's the thing: the dtype attributes of both ndarray and _array_api.Array currently return np.dtype instances,
so neither of them can currently be (validly) used as x in this example.

I'm afraid I'm not following you here. Are you saying the dtype itself should have a dtype attribute, like `int8.dtype`?

This is how this module currently works

```py
>>> import numpy._array_api as npa
>>> x = npa.ones((2, 3), dtype=npa.int8)
>>> x.dtype is npa.int8
True
```

Note that I'm using instances for `npa.int8` specifically so that they are exactly the same thing returned by `x.dtype` for any array type. Previously `npa.int8` was the dtype class (same as `np.int8`), but this lead to some problems due to the way equality testing on NumPy dtypes work, and based on the comments here, it would presumably also cause some issues with the type hints.

The spec really requires nothing from the dtype objects except that they be comparable with equality. I had previously suggested to make custom classes for the dtypes in the array API namespace to make this more explicit. Right now, `np._array_api.float64` has a lot of attributes that aren't part of the array API spec. But this idea was rejected (by @rgommers at least, I don't remember if others have weighed in on the idea).